compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -944,8 +943,31 @@ public void visit(BLangRecordLiteral recordLiteral) {
         }
     }
 
-    private List<BType> getRecordCompatibleType(BType bType, BLangRecordLiteral recordLiteral) {
+    private boolean isMappingConstructorCompatibleType(BType type) {
+        return type.tag == TypeTags.RECORD || type.tag == TypeTags.MAP || type.tag == TypeTags.JSON;
+    }
+
+    private void reportIncompatibleMappingConstructorError(BLangRecordLiteral mappingConstructorExpr, BType expType) {
```

Comment:
Should we special case the `T?` case where T is a record? If so, can provide an even better error message for the particular issue mentioned in #17360

Reply:
+1, `T?` is 80% of the case, when considering unions.